doc/new_architecture.md

```diff
@@ -0,0 +1,161 @@
+* Implement a higher-level API to be consumed by the clients, partially replacing the D-Bus.
+
+### Components
+
+With those goals in mind, we are considering the following components:
+
+* **Agama core (old `agama-dbus-server`)**: implements the installation logic. It relies heavily on
+the Agama YaST service.
+
+* **Agama YaST service (old `rubygem-agama`)**: it is written in Ruby and has direct access to YaST
+libraries. Complex parts, like storage and software handling, are implemented in this component.
+
+* **HTTP and WebSocket API**: implements the API the clients should use to communicate with Agama.
+
+* **Web user interface (old `cockpit-agama`)**: Agama's graphical user interface. The web server
+makes this React application available to the browsers.
+
+* **Command Line Interface (`agama-cli`)**: it allows interaction with Agama core and drives the
+auto-installation process. With the new architecture, connecting through the network might be
+possible without SSH.
+
+The following diagram could be better, but it represents the main components and their interactions.
+
+```mermaid
+flowchart LR
+  subgraph Clients
+    Browser
+    CLI
+  end
+
+  subgraph Agama Core
+    subgraph Web["Web Server"]
+      direction TB
+      UI["Web UI"]
+      API["HTTP/WS API"]
+      WS["WebSocket"]
+    end
+
+    Web <--"Channel?" --> Rust
+  end
+  Rust <-- "D-Bus" --> YaST["Agama YaST"]
+
+  Browser --> UI
+  Browser --> API
+  Browser <---> WS
+  CLI --> API
+  CLI <---> WS
+```
+
+### The web-based API
+
+The new web-based API is divided into two different parts:
+
+* **HTTP/JSON API**: allows clients to execute actions or query Agama. For instance, it could get
+the list of available products, request a new storage proposal or start the installation. About the
+approach, something REST-like is the most suitable. Switching to GraphQL or gRPC do not seem to be
+needed in our case (todo: write why).
+
+* **WebSocket**: Agama will use WebSockets to notify clients about any relevant event: progress,
+configuration changes, etc. The event's format is not defined yet, but a JSON event containing the
+`type` and the `payload/details` should be enough[^topics].
+
+[^topics]: Ideally, the clients should be able to subscribe to the topics they are interested in.
+    But that feature can wait.
+
+### Encryption
+
+In the case of a remote installation, the communication between the clients and the server must be
+encrypted. Connecting to port 80 (HTTP) should redirect the client to port 443 (HTTPS).
+
+About the certificate, Agama will use a self-signed certificate unless the user injects its own
+certificate (through a kernel command line option).
+
+NOTE: under discussion.
+
+### Authentication
+
+The HTTP interface should allow authentication specifying a user and password that will be checked
+against PAM. It is not clear yet, but we might need to check whether the logged user has permissions
+(most probably through Policy Kit).
```

Comment:
Do we need to allow non root user for installation? Does it make sense at all?

Reply:
No, I do not think so. But I wondered whether the right question is: "do you have permissions?" instead of "are you root?". However, I only have a little experience with Polkit (which reminds me that the correct name is *Polkit*).

Reply:
Well, as user of installation you can do almost anything to system. So using just root makes it easier from my POV. For me better question is if we want to allow e.g. openssh keys instead of root password and enforce it. So customer can disable root password and enforce usage of ssh keys.

Reply:
In my opinion, that's a different (but related) topic. Anyway, we can keep the *most probably through Polkit* text and re-evaluate before implementing.

Reply:
And, yes, absolutely we should allow the users to inject their own SSH public keys.